File: platform/core/ui/static/js/flater/datakvalitet.js
// Datakvalitet (M-3, 092) — profilen av plattformens egne tenant-tabeller.
//
// FLATEN VISER, DEN REGNER IKKE. Hvert tall her står i en rad API-et
// leverte: `rader_vurdert`, `rader_avvik` og `andel_avvik` kommer alle
// tre fra basen, og andelen er en GENERERT kolonne der — flaten deler
// aldri to av svarets tall på hverandre (M-16-regelen). Det eneste som
// skjer under er presentasjon: en andel i [0, 1] vises som prosent med
// én desimal, og en maskinkode slås opp som setning.
//
// AVVIKSANDEL SOM TALL OG TEKST, ALDRI TRAFIKKLYS ALENE. Hver rad viser
// «3 av 128 (2,3 %)» i sin egen celle. Et fargefelt kunne stått ved
// siden av; det kan ikke stå i stedet for (WCAG 1.4.1) — og en andel
// uten sine tellere er et tall ingen kan etterprøve.
//
// «IKKE MÅLT» ER IKKE «0». Modulens bærende regel, rendret: en regel som
// står i kjøringens `umaalbare_regler` får teksten «ikke målt», ikke en
// tom celle og ikke en null. En regel uten profilrad som HELLER ikke er
// umålbar, har ingen rader hos denne tenanten — og det er en tredje
// tilstand med sin egen setning. Tre tilstander, tre setninger; ingen av
// dem er en tom celle.
//
// KJØRINGENS HODE SIER `avbrutt` MED TEKST. En runde som ikke rakk
// gjennom registeret er ikke en grønn runde med få regler, og det står
// som en setning i hodet — ikke som et ikon, og ikke bare som et tall.
//
// TABELLENE ER EKTE TABELLER: <caption>, th scope="col" og en
// th scope="row" som navngir hver rad. `aria-sort` står på kolonnen
// SERVEREN sorterte på — dørene i 092 leverer regler alfabetisk og funn
// nyeste først — og den er derfor sann uten at flaten sorterer om. Det
// er også grunnen til at det ikke finnes en sorteringsknapp her: en
// `aria-sort` som lyver er verre enn ingen.
//
// INGEN KNAPPER. Profileringen skrives av
// `disponit-kvalitetsprofil.service` med sin egen DB-rolle, og
// kvalitetsregisteret endres kun i migrasjon. Det finnes ingen HTTP-dør
// å tegne en knapp til — og v1 RETTER ingenting, slår ingenting sammen
// og blokkerer ingen bestilling. Det er en dom, ikke en manglende
// funksjon, og flaten skal ikke antyde noe annet.
import { el, sett } from "../dom.js";
import { t, sprak } from "../i18n.js";
import { hentJson } from "../api.js";
import { Tidspunkt } from "../komponenter.js";
import { flateHode, medStatus } from "./felles.js";

// Maskinkoder → setninger. Reserven er koden selv: en femte funntype
// eller en ny regeltype skal bli SYNLIG som en ukjent kode, ikke falle
// ut av tabellen.
function funntypeTekst(kode) {
  return t(`ui.datakvalitet.funntype.${kode}`, kode);
}
function regeltypeTekst(kode) {
  return t(`ui.datakvalitet.regeltype.${kode}`, kode);
}
function alvorlighetTekst(kode) {
  return t(`ui.datakvalitet.alvorlighet.${kode}`, kode);
}

// Andel i [0, 1] → prosent med én desimal. EN verdi omregnet til en
// annen enhet, ikke et forhold mellom to av svarets tall.
//
// Desimalskilletegnet er LOCALENS, ikke en hardkodet komma (CodeRabbit):
// «2,3 %» er riktig på norsk og feil på engelsk, og et tall som leses
// feil er verre enn intet tall. `Intl.NumberFormat` gjør både skilletegn,
// tusenskille og prosenttegnets plassering til språkets ansvar.
function prosent(andel) {
  if (typeof andel !== "number") return "—";
  return new Intl.NumberFormat(sprak(), {
    style: "percent", minimumFractionDigits: 1, maximumFractionDigits: 1,
  }).format(andel);
}

// Avvikscellen: tallene FØRST, andelen etter. Leses «3 av 128 (2,3 %)».
function avvikTekst(p) {
  return t("ui.datakvalitet.avvik")
    .replace("{avvik}", String(p.rader_avvik))
    .replace("{vurdert}", String(p.rader_vurdert))
    .replace("{andel}", prosent(p.andel_avvik));
}

function regelseksjon(regler) {
  const seksjon = el("section", { class: "kpi-kort" },
    el("h3", { text: t("ui.datakvalitet.regler.tittel") }));
  if (!regler.length) {
    seksjon.append(el("p", { class: "muted",
      text: t("ui.datakvalitet.regler.ingen") }));
    return seksjon;
  }
  const tabell = el("table", { class: "kpi-tabell" },
    el("caption", { text: t("ui.datakvalitet.regler.caption") }));
  tabell.append(el("thead", {}, el("tr", {},
    // Dørens rekkefølge ER alfabetisk på regel-ID (092). aria-sort sier
    // det, og flaten sorterer ikke om.
    el("th", { scope: "col", "aria-sort": "ascending",
      text: t("ui.datakvalitet.regler.kolonne.regel") }),
    el("th", { scope: "col",
      text: t("ui.datakvalitet.regler.kolonne.maal") }),
    el("th", { scope: "col",
      text: t("ui.datakvalitet.regler.kolonne.type") }),
    el("th", { scope: "col",
      text: t("ui.datakvalitet.regler.kolonne.alvorlighet") }),
    el("th", { scope: "col",
      text: t("ui.datakvalitet.regler.kolonne.begrunnelse") }))));
  const tbody = el("tbody");
  for (const r of regler) {
    tbody.append(el("tr", {},
      el("th", { scope: "row", text: r.regel_id }),
      el("td", { text: `${r.relasjon}.${r.kolonne}` }),
      el("td", { text: regeltypeTekst(r.regeltype) }),
      el("td", { text: alvorlighetTekst(r.alvorlighet) }),
      el("td", { class: "celle-tekst", text: r.begrunnelse })));
  }
  tabell.append(tbody);
  seksjon.append(tabell);
  return seksjon;
}

// Hodet for siste kjøring. Fire tall og TO setninger: én om når, én om
// hvorvidt runden rakk gjennom registeret.
function kjoringshode(k) {
  const hode = el("div", {});
  const naar = el("p", {});
  sett(naar, t("ui.datakvalitet.kjoring.naar"), " ",
    Tidspunkt(k.startet_ts, {}));
  const tall = el("p", {
    text: t("ui.datakvalitet.kjoring.tall")
      .replace("{regler}", String(k.antall_regler))
      .replace("{umaalbare}", String(k.antall_umaalbare))
      .replace("{funn}", String(k.antall_funn)),
  });
  // `avbrutt` SOM TEKST. Den fullførte runden sier det også — ellers
  // ville fraværet av en advarsel vært det eneste signalet, og et
  // fravær er ikke en melding.
  const dom = el("p", {
    // Ingen egen fargeklasse: dommen ER teksten. En klasse som bare
    // fantes for å farge den ville vært trafikklyset denne flaten
    // nekter å bruke, og «avbrutt» må uansett leses for å forstås.
    class: "muted",
    text: k.avbrutt
      ? t("ui.datakvalitet.kjoring.avbrutt")
      : t("ui.datakvalitet.kjoring.fullfort"),
  });
  hode.append(naar, tall, dom);
  return hode;
}

// Profiltabellen for ÉN kjøring, satt sammen med registeret: hver regel
// får en rad, også de uten profiltall. Det er hele poenget — en regel
// som mangler skal SI hvorfor den mangler.
function profiltabell(k, regler) {
  const perRegel = new Map();
  for (const p of k.profiler || []) perRegel.set(p.regel_id, p);
  const umaalbare = new Set(k.umaalbare_regler || []);
  const tabell = el("table", { class: "kpi-tabell" },
    el("caption", { text: t("ui.datakvalitet.profil.caption") }));
  tabell.append(el("thead", {}, el("tr", {},
    el("th", { scope: "col", "aria-sort": "ascending",
      text: t("ui.datakvalitet.profil.kolonne.regel") }),
    el("th", { scope: "col",
      text: t("ui.datakvalitet.profil.kolonne.vurdert") }),
    el("th", { scope: "col",
      text: t("ui.datakvalitet.profil.kolonne.avvik") }))));
  const tbody = el("tbody");
  for (const r of regler) {
    const p = perRegel.get(r.regel_id);
    let vurdert;
    let avvik;
    if (umaalbare.has(r.regel_id)) {
      // MODULENS BÆRENDE REGEL PÅ SKJERMEN: en regel som ikke kunne
      // måles står som «ikke målt», aldri som 0.
      vurdert = t("ui.datakvalitet.profil.ikke_maalt");
      avvik = t("ui.datakvalitet.profil.ikke_maalt");
    } else if (!p) {
      // Målt, men tenanten har ingen rader i relasjonen. En tredje
      // tilstand, og ikke det samme som «ikke målt».
      vurdert = t("ui.datakvalitet.profil.ingen_rader");
      avvik = t("ui.datakvalitet.profil.ingen_rader");
    } else {
      vurdert = String(p.rader_vurdert);
      avvik = avvikTekst(p);
    }
    tbody.append(el("tr", {},
      el("th", { scope: "row", text: r.regel_id }),
      el("td", { text: vurdert }),
      el("td", { text: avvik })));
  }
  tabell.append(tbody);
  return tabell;
}

function profilseksjon(d) {
  const seksjon = el("section", { class: "kpi-kort" },
    el("h3", { text: t("ui.datakvalitet.profil.tittel") }));
  const kjoringer = (d && d.kjoringer) || [];
  if (!kjoringer.length) {
    // ÆRLIG TOMTILSTAND: ingen profilering er ikke «alt i orden». Det er
    // en base ingen har målt, og setningen sier det.
    seksjon.append(el("p", { class: "muted",
      text: t("ui.datakvalitet.profil.ingen") }));
    return seksjon;
  }
  // Nyeste først er dørens rekkefølge (092) — flaten sorterer ikke om.
  const siste = kjoringer[0];
  seksjon.append(kjoringshode(siste),
    profiltabell(siste, (d && d.regler) || []));
  return seksjon;
}

function funntabell(funn, medTenant) {
  const tabell = el("table", { class: "kpi-tabell" },
    el("caption", { text: medTenant
      ? t("ui.datakvalitet.funn.caption_tverrgaaende")
      : t("ui.datakvalitet.funn.caption") }));
  const hoderad = el("tr", {},
    // Dørens rekkefølge er `sist_sett_ts DESC` (092).
    el("th", { scope: "col", "aria-sort": "descending",
      text: t("ui.datakvalitet.funn.kolonne.sist_sett") }));
  if (medTenant) {
    hoderad.append(el("th", { scope: "col",
      text: t("ui.datakvalitet.funn.kolonne.tenant") }));
  }
  hoderad.append(
    el("th", { scope: "col", text: t("ui.datakvalitet.funn.kolonne.regel") }),
    el("th", { scope: "col", text: t("ui.datakvalitet.funn.kolonne.type") }),
    el("th", { scope: "col",
      text: t("ui.datakvalitet.funn.kolonne.ganger") }));
  tabell.append(el("thead", {}, hoderad));
  const tbody = el("tbody");
  for (const f of funn) {
    const rad = el("tr", {},
      el("th", { scope: "row" }, Tidspunkt(f.sist_sett_ts, {})));
    if (medTenant) rad.append(el("td", { text: f.tenant }));
    rad.append(
      el("td", { text: f.regel_id }),
      el("td", { text: funntypeTekst(f.funntype) }),
      el("td", { text: String(f.ganger_sett) }));
    tbody.append(rad);
  }
  tabell.append(tbody);
  return tabell;
}

function funnseksjon(d) {
  const seksjon = el("section", { class: "kpi-kort" },
    el("h3", { text: t("ui.datakvalitet.funn.tittel") }));
  const funn = (d && d.funn) || [];
  if (!funn.length) {
    seksjon.append(el("p", { class: "muted",
      text: t("ui.datakvalitet.funn.ingen") }));
  } else {
    seksjon.append(funntabell(funn, false));
  }
  return seksjon;
}

// Plattformdriftens seksjon. Den tegnes KUN når serveren sa
// `plattformdrift: true` — og den SIER at den er tverrgående, fordi en
// tabell som stille inneholder andre kunders rader er verre enn en som
// forklarer hvorfor den gjør det.
function tverrgaaendeSeksjon(d) {
  if (!d || !d.plattformdrift) return null;
  const seksjon = el("section", { class: "kpi-kort" },
    el("h3", { text: t("ui.datakvalitet.tverrgaaende.tittel") }),
    el("p", { class: "muted",
      text: t("ui.datakvalitet.tverrgaaende.forklaring") }));
  const funn = d.tverrgaaende_funn || [];
  if (!funn.length) {
    seksjon.append(el("p", { class: "muted",
      text: t("ui.datakvalitet.tverrgaaende.ingen") }));
  } else {
    seksjon.append(funntabell(funn, true));
  }
  return seksjon;
}

export function visDatakvalitet(hoved, ctx) {
  sett(hoved, ...flateHode(t("ui.datakvalitet.tittel"),
    t("ui.datakvalitet.undertittel")));
  const kropp = el("div", { class: "kpi-kort-liste" });
  hoved.append(kropp);
  medStatus(hoved, ctx,
    () => hentJson("/v1/datakvalitet"),
    (data) => {
      sett(hoved, ...flateHode(t("ui.datakvalitet.tittel"),
        t("ui.datakvalitet.undertittel")), kropp);
      const seksjoner = [profilseksjon(data), funnseksjon(data),
                         regelseksjon((data && data.regler) || [])];
      const tverr = tverrgaaendeSeksjon(data);
      if (tverr) seksjoner.push(tverr);
      sett(kropp, ...seksjoner);
    });
}

File: platform/core/api/datakvalitet.py
"""M-3 datakvalitet — leseflatens data (092).

AVLESNING, ALDRI ANALYSE. Modulen utleder ingenting og regner ingenting
om: hvert tall her står i en rad `m3_profiler` alt har committet, og den
eneste formingen som skjer er GRUPPERINGEN av dørens flate rader
(kjøring × regel) til én kjøring med sine profiltall — presentasjon, som
etter oversikt-lærdommen bor i API-laget og ikke i SQL.

TO AUTORITETER PÅ SAMME RUTE (`/v1/utrulling`-presedensen):

  * `security:read` — kundens/ops-øktens egen tilstand: registeret (som
    er globalt og uten tenantdata), kjøringenes hoder (plattformskop:
    fire tall om målingen selv) og TENANTENS EGNE profiltall og funn.
    Isolasjonen er `tenant_isolasjon` i basen, ikke et filter her.
  * `platform:admin` — plattformdriften ser I TILLEGG funnlisten på
    tvers av tenanter. Utvidelsen avgjøres HER, inne i endepunktet, av
    samme grunn som utrullingsplanen: det er en utvidelse av svaret, ikke
    en annen inngang. Scopet er ikke i `LESESCOPES`, så en browsersesjon
    når det aldri (verifisert i `app.py::_autentiser`) — og det er
    riktig: dette er en maskin-/ops-autoritet.

DEN BÆRENDE REGELEN FØLGER MED UT. `umaalbare_regler` sendes videre per
kjøring, slik at flaten kan si «ikke målt» ved riktig regel i stedet for
å tegne en tom celle som ser ut som null avvik. Uten det ville
«0 tomme felt» fordi målingen ikke kjørte sett ut som en grønn profil på
skjermen — nøyaktig løgnen modulen finnes for å ikke fortelle.

INGEN MUTASJON FINNES SOM HTTP. Profileringen skrives av
`disponit-kvalitetsprofil.service` med rollen `disponit_kvalitetsmaaler`,
en fullmakt web-API-et ikke har og ikke skal ha, og registeret endres
kun i migrasjon.
"""
from __future__ import annotations

#: Plattformdriftens autoritet. Skilt fra `security:read`, som er en
#: TENANTBUNDET ops/compliance-scope på en kundesesjon (PR-008 §1) — den
#: sier ingenting om rett til å se andre kunders funn.
PLATTFORMDRIFT = "platform:admin"

#: Taket på hvor mange KJØRINGER flaten henter. Dørene klemmer selv til
#: sine egne spenn; tallet her er standarden når klienten ikke ber om noe.
#: Dette er en AVLESNING, ikke en paginert logg.
GRENSE_STANDARD = 10
#: Taket på funnlister. Funnlisten vokser ikke med kadensen (ett funn per
#: regel/tenant/funntype), så den er kort per konstruksjon.
FUNNGRENSE = 200


def _ts(verdi) -> str | None:
    return verdi.isoformat() if verdi is not None else None


def _andel(verdi) -> float | None:
    """`numeric` kommer tilbake som Decimal og er ikke JSON-serialiserbar.

    `float` er riktig her og skjuler ingenting: dette er en ANDEL i
    [0, 1] regnet av basen som en generert kolonne, ikke et beløp. Den
    følges alltid av sine to tellere (`rader_vurdert`, `rader_avvik`), så
    flaten kan vise tallet og ikke bare fargen.
    """
    return None if verdi is None else float(verdi)


def regler(conn, tenant: str) -> list[dict]:
    """Registeret. Globalt, uten tenantdata — men bak tenantkontekst,
    fordi RETTEN til å spørre er øktens selv når dataene ikke er det."""
    rader = conn.execute("SELECT * FROM m3_regelregister(%s)",
                         (tenant,)).fetchall()
    return [{"regel_id": r[0], "relasjon": r[1], "kolonne": r[2],
             "regeltype": r[3], "uttrykk": r[4], "alvorlighet": r[5],
             "terskel_andel": _andel(r[6]), "begrunnelse": r[7]}
            for r in rader]


def kjoringer(conn, tenant: str, grense: int = GRENSE_STANDARD) -> list[dict]:
    """Siste kjøringer med tenantens EGNE profiltall.

    Døren gir FLATE rader (kjøring × regel) fordi grupperingen er
    presentasjon. Den skjer her, og den bevarer dørens rekkefølge: nyeste
    kjøring først, reglene alfabetisk innenfor hver. En kjøring UTEN
    profilrader for denne tenanten står fortsatt i listen — den skjedde,
    og hodet er den ærligste delen av svaret.
    """
    rader = conn.execute("SELECT * FROM m3_kvalitetsprofil(%s,%s)",
                         (tenant, grense)).fetchall()
    ut: list[dict] = []
    indeks: dict[str, dict] = {}
    for (kid, startet, fullfort, n_regler, n_umaalbare, n_funn,
         umaalbare, avbrutt, alder_s, regel_id, vurdert, avvik,
         andel) in rader:
        nokkel = str(kid)
        k = indeks.get(nokkel)
        if k is None:
            k = {"kjoring_id": nokkel, "startet_ts": _ts(startet),
                 "fullfort_ts": _ts(fullfort), "alder_s": alder_s,
                 "antall_regler": n_regler,
                 "antall_umaalbare": n_umaalbare,
                 "antall_funn": n_funn,
                 # Navnene, ikke bare tallet: flaten skal kunne si «ikke
                 # målt» ved RIKTIG regel.
                 "umaalbare_regler": list(umaalbare or ()),
                 "avbrutt": bool(avbrutt),
                 "profiler": []}
            indeks[nokkel] = k
            ut.append(k)
        # LEFT JOIN-en gir én rad med NULL-regel for en kjøring uten
        # tenantrader. Den er ikke en profil og skal ikke bli en tom rad
        # i tabellen.
        if regel_id is not None:
            k["profiler"].append({
                "regel_id": regel_id, "rader_vurdert": vurdert,
                "rader_avvik": avvik, "andel_avvik": _andel(andel)})
    return ut


def egne_funn(conn, tenant: str, grense: int = FUNNGRENSE) -> list[dict]:
    """Tenantens egne funn. Ren RLS — ingen tenantkolonne i svaret,
    fordi det bare finnes én tenant å snakke om her."""
    rader = conn.execute("SELECT * FROM m3_kvalitetsfunn(%s,%s)",
                         (tenant, grense)).fetchall()
    return [{"regel_id": r[0], "funntype": r[1], "forst_sett_ts": _ts(r[2]),
             "sist_sett_ts": _ts(r[3]), "ganger_sett": r[4],
             "detaljer": r[5] if isinstance(r[5], dict) else {}}
            for r in rader]


def tverrgaaende_funn(conn, tenant: str, grense: int = FUNNGRENSE) -> list[dict]:
    """Funnlisten på tvers — KUN for `platform:admin`.

    Kalles bare når endepunktet har avgjort at økten har
    plattformdriftens autoritet. Døren åpner sitt kryss-tenant-vindu
    lokalt i transaksjonen og lukker det med den; policyen er
    `FOR SELECT`, så vinduet kan ikke bli en skrivevei uansett.
    """
    rader = conn.execute(
        "SELECT * FROM m3_kvalitetsfunn_tverrgaaende(%s,%s)",
        (tenant, grense)).fetchall()
    return [{"tenant": r[0], "regel_id": r[1], "funntype": r[2],
             "forst_sett_ts": _ts(r[3]), "sist_sett_ts": _ts(r[4]),
             "ganger_sett": r[5],
             "detaljer": r[6] if isinstance(r[6], dict) else {}}
            for r in rader]


def svar(conn, tenant: str, scopes, grense: int = GRENSE_STANDARD) -> dict:
    """Hele flatens data i ett svar.

    `plattformdrift` står i svaret fordi flaten må kunne SI at den viser
    en tverrgående liste — en tabell som stille inneholder andre kunders
    rader er verre enn en som sier hvorfor den gjør det.
    """
    plattformdrift = PLATTFORMDRIFT in set(scopes or ())
    ut = {
        "plattformdrift": plattformdrift,
        "regler": regler(conn, tenant),
        "kjoringer": kjoringer(conn, tenant, grense),
        "funn": egne_funn(conn, tenant),
    }
    # Uten plattformdrift finnes NØKKELEN ikke — ikke en tom liste. En
    # tom liste ville lest som «ingen funn på tvers», og det er ikke det
    # svaret sier.
    if plattformdrift:
        ut["tverrgaaende_funn"] = tverrgaaende_funn(conn, tenant)
    return ut
